libs/core/device-manager/observers/usb-device-detection.observer.ts

```diff
@@ -0,0 +1,69 @@
+/**
+ * Copyright (c) Mudita sp. z o.o. All rights reserved.
+ * For licensing, see https://github.com/mudita/mudita-center/blob/master/LICENSE.md
+ */
+
+import { Observer } from "Core/core/types"
+import { DeviceManager } from "Core/device-manager/services"
+import { ipcMain } from "electron-better-ipc"
+import { ApiSerialPortToRendererEvents } from "device/models"
+import { SerialPortDeviceAdapterEvent } from "Core/device/modules/mudita-os/adapters"
+
+const intervalTime = 3000
+
+export class UsbDeviceDetectionObserver implements Observer {
+  private previousAttachedDevicePaths = new Set<string>()
+  constructor(private deviceManager: DeviceManager) {}
+
+  public observe(): void {
+    void this.watchAttachedDevices()
+    this.mountListeners()
+  }
+
+  private async watchAttachedDevices(): Promise<void> {
+    await this.detectDeviceStateChange()
+
+    return new Promise<void>((resolve) => {
+      setTimeout(() => {
+        void (async () => {
+          resolve(await this.watchAttachedDevices())
+        })()
+      }, intervalTime)
+    })
+  }
+
+  private mountListeners() {
+    ipcMain.on(ApiSerialPortToRendererEvents.Closed, () => {
+      void this.detectDeviceStateChange()
+    });
+    ipcMain.on(SerialPortDeviceAdapterEvent.Closed, () => {
+      void this.detectDeviceStateChange()
+    });
+  }
+
+  private async detectDeviceStateChange(): Promise<void>  {
+    const attachedDevices = await this.deviceManager.getAttachedDevices()
+    const detachedDevicePaths = Array.from(
+      this.previousAttachedDevicePaths
+    ).filter(
+      (previousAttachedDevicePath) =>
+        !attachedDevices
+          .map(({ path }) => path)
+          .includes(previousAttachedDevicePath)
+    )
+
+    detachedDevicePaths.forEach((detachedDevicePath) => {
+      void this.deviceManager.removeDevice(detachedDevicePath)
+    })
+
+    attachedDevices.forEach((attachedDevice) => {
+      if (!this.previousAttachedDevicePaths.has(attachedDevice.path)) {
+        void this.deviceManager.addDevice(attachedDevice)
```

Comment:
Within the context of this service, we cannot be certain whether the appearance of a new device with the same path is an error. Nevertheless, there is a short time window due to asynchronous operations that may result in the emergence of a new device instead of the old one with the same path.

In such a situation, the application attempts to make a request, and the logic responsible for device initialization redirects to the troubleshooting view.